bin/target.ml

```diff
@@ -106,9 +106,41 @@ let resolve_path path ~(setup : Dune.Main.build_system) =
 let resolve_target common ~(setup : Dune.Main.build_system) s =
   match Alias.of_string common s ~contexts:setup.workspace.contexts with
   | Some a -> Ok [Alias a]
-  | None ->
-    let path = Path.relative Path.root (Common.prefix_target common s) in
-    resolve_path path ~setup
+  | None -> begin
+      let dep =
+        let parser =
+          Dune.Syntax.set Dune.Stanza.syntax Dune.Stanza.latest_version
+            Dune.Dune_file.Dep_conf.decode
+        in
+        Dune_lang.Decoder.parse parser Univ_map.empty
+          (Dune_lang.parse_string ~fname:"command line" ~mode:Dune_lang.Parser.Mode.Single s)
+      in
+      match dep with
+      | Dune.Dune_file.Dep_conf.File sw ->
+        let f ctx =
+          let sctx = String.Map.find_exn setup.scontexts (Context.name ctx) in
+          let dir =
+            Path.Build.relative ctx.Context.build_dir
+              (String.concat ~sep:Filename.dir_sep common.root.to_cwd)
+          in
+          let expander = Dune.Super_context.expander sctx ~dir in
+          let lookup_module ~dir name =
+            Dune.Dir_contents.Dir_artifacts.lookup_module
+              (Dune.Dir_contents.artifacts (Dune.Dir_contents.get sctx ~dir)) name
+          in
+          let lookup_library ~dir name =
+            Dune.Dir_contents.Dir_artifacts.lookup_library
+              (Dune.Dir_contents.artifacts (Dune.Dir_contents.get sctx ~dir)) name
+          in
+          let expander = Dune.Expander.set_lookup_module expander ~lookup_module in
+          let expander = Dune.Expander.set_lookup_library expander ~lookup_library in
```

Comment:
It would be better to set these lookup functions inside the function `Super_context.expander` but `Dir_contents` depends on `Super_context` so there is a dependency issue.